### DISCUSSION TITLE
How to use Julia engine when qmd uses packages?

### DISCUSSION THREAD

**Discussion by sbacelar** [quarto-dev/quarto-cli#9712] [Q&A]
### Description

  I am trying to use Julia with the Julia engine  QuartoNotebookRunner.jl.
  I have installed Quarto pre-release build 1.5.37 for mac. I have been able to run the simple example on the terminal and rendering a pdf.
  
  but not the one in the same web page (https://quarto.org/docs/computations/julia.html) where there is a need for the `Plots` package. I added the Plots package to my work directory and tried to do `render document.qmd --to pdf` and I got this error insisting that I should install `Plots`:
  ```
  shell> quarto render document.qmd --to pdf
  Running [1/1] at line 16:  using Plots
  ERROR: Julia server returned error after receiving "run" command:
  Failed to run notebook: /Users/sergiobacelar/Dropbox/JULIA/sbquarto_2/document.qmd
 ERROR: EvaluationError: Encountered 1 error during evaluation
 
 Error 1 of 1
 @ /Users/sergiobacelar/Dropbox/JULIA/sbquarto_2/document.qmd:16
 ArgumentError: Package Plots not found in current path.
  - Run `import Pkg; Pkg.add("Plots")` to install the Plots package.```

**Comment by cscheid** [comment]
@jkrumbiegel Would you mind helping out here?

**Reply by mcanouil** [reply]
The issue is a documentation issue, in particular "Clarify that the Julia engine uses project environment which needs to be setup by the users before running Quarto: Pkg.activate(".")" from

- https://github.com/quarto-dev/quarto-cli/issues/9561

Same report as here:

- https://github.com/quarto-dev/quarto-cli/discussions/9587
or 
- https://github.com/quarto-dev/quarto-cli/discussions/9551

In all cases, it's all about Julia environment that needs to be activated similarly to Virtual Env for Python.
The documentation about that is: https://pkgdocs.julialang.org/v1/environments/

For now the documentation does not talk at all about environment and uses keyboard shortcuts instead of actual commands.
<img width="890" alt="Screenshot 2024-05-18 at 18 19 14" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/339719ba-95f7-4ec9-a7f9-89e4cff5eb81">

In the screenshot, `]` is not a character "to type" as in "to write" but to press the corresponding keystrokes, then you actually type `activate`.

https://github.com/quarto-dev/quarto-cli/assets/8896044/adf5ab0b-8cc9-4b09-922b-7fd4f6548a9f

**Reply by cderv** [reply]
I guess we have a very different logic currently in Julia engine compared to others as we do this:
> When you first attempt to render a notebook with the julia engine, Quarto will automatically install this package into a private environment that is owned by Quarto. This installation will therefore not interact or conflict with any other Julia environments on your machine

So we create an environment used by default, and currently we don't really document how to add to this specific environment used by Quarto.  So we rely for now on the documentation inside yellow box for using a custom QuartoNotebookRunner version, provided in a custom env. 

We do differently for Jupyter or Knitr. Either we find Python and Jupyter , or R and rmarkdown, either we error. And if we found it, we used the associated environment (libraries) without creating a specific one. We rely on current PATH configuration, or other tools (like **renv** in R or **virtualenv** in Python). 

I wonder if there is good way to install some packages in this Quarto specific Julia environment. Otherwise, maybe this should not be done this way and QuartoNotebookRunner not provided with Quarto. We could just document that
- QuartoNotebookRunner needs to be installed and available in the environment found (same as we do for rmarkdown/knitr and jupyter)
- document how to change the environment used by Quarto setting the environment var, the `exeFlags` or maybe a new YAML config.
- Possibly integrate with Project.Toml more closely, using this project environment by default, if file is present at project root. 

Anyhow, I think this needs to be documented for sure, but possibly improve at user experience level considering the different feedback we have. 

@jkrumbiegel having your view on how you thought this would be used would be really helpful !

**Reply by jkrumbiegel** [reply]
There's a misunderstanding here I think:

> currently we don't really document how to add to this specific environment used by Quarto

You're not supposed to install anything into this environment because it just contains the resolved list of packages needed to run QuartoNotebookRunner.jl itself. This is decoupled from the packages a user might want to load in a given qmd file.

The logic goes like this: Quarto installs QuartoNotebookRunner into its own private environment. When quarto renders a qmd file with the julia engine, it sends that file to a server process running QuartoNotebookRunner. This server process then starts or reuses a separate Julia worker process for each qmd file. In _this_ Julia process, the user may activate a certain environment if they wish to do so. _Not_ in the server process which relates to the private environment.

We chose to default to using a qmd file's directory as the default environment for its worker process. In the case where the user never activates and manipulates an environment in this folder, this is equivalent to there being an empty environment there, which means that only stdlib packages can be loaded. But this is actually good, because if we were to load the shared environment that a normal Julia process would usually load, let's say `@v1.10` if we're running Julia 1.10, then the users would be nudged towards a workflow where their notebooks end up not being reproducible. If you run notebook A off the main shared env, then a week later a new notebook B, but you install new packages into the main shared env for B, then you are in danger of breaking A. This gets worse the more notebooks and the more packages you have.

Our workflow signals to the user "hey, you have no packages but stdlib available, create an environment file if you want to change this". I completely agree with @mcanouil that this could be documented better for those trying out Julia for the first time. I'm so used to this workflow that I probably left out too many details when writing the docs. For example the `]activate` shorthand will be immediately recognizable to most Julia users as relating to the Pkg REPL mode, but it throws off people copy-pasting this code.

**Reply by cscheid** [reply]
@jkrumbiegel Thanks, this is very helpful. I think that adding step-by-step instructions to our docs explaining how to install (say) `Plots` in one such environment would be a big improvement, and wouldn't take a lot of additional space.

**Reply by cderv** [reply]
Thanks a lot @jkrumbiegel ! This clarifies a lot - we should definitely add more about this in our doc.

**Reply by jkrumbiegel** [reply]
I have written up something in https://github.com/quarto-dev/quarto-web/pull/1102

**Reply by sbacelar** [reply]
I usually use Quarto with R. Now, I am trying to use it with Julia, and I am having some problems. I only tried to run the example on the web page [https://quarto.org/docs/computations/julia.html] using the Jupyter and Julia engines and Quarto version 1.5.37. Let me say first that I had no problems running it on Ubuntu (I had only to install `rsvg-convert`) and partially on WSL2. But I had problems running the same example on MacOS (GFK and Plots not found) and Windows 11 (I couldn't install `rsvg-convert` with `Chocolatey`, and I also had problems with Plots. I am including the error messages I received in each case:

## MacOS 11.7.10 jupyter engine: quarto render --to pdf

Outputs pdf but inserts the following after the figure:


`GKS: dlopen(/Users/sergiobacelar/.julia/artifacts/86a2153e203fb861d380988509048a2ffa8e3a9f/li`
  `Referenced from: /Users/sergiobacelar/.julia/artifacts/86a2153e203fb861d380988509048a2ffa8e`
  `Reason: Incompatible library version: cairoplugin.so requires version 11603.0.0 or later, b`

## MacOS 11.7.10 julia engine: quarto render --to pdf

(sbquarto_2) pkg> st
Status `~/Dropbox/JULIA/sbquarto_2/Project.toml`
  [91a5bcdd] Plots v1.40.4

shell> quarto render document.qmd --to pdf
Running [1/1] at line 16:  using Plots
ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: /Users/sergiobacelar/Dropbox/JULIA/sbquarto_2/document.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ /Users/sergiobacelar/Dropbox/JULIA/sbquarto_2/document.qmd:16
ArgumentError: Package Plots not found in current path.
- Run `import Pkg; Pkg.add("Plots")` to install the Plots package.

## Windows 11 julia engine: quarto preview --to pdf

Julia server process started.
Running [1/1] at line 16:  using Plots
ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: D:\Dropbox\JULIA\sbquarto_2\document.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ D:\Dropbox\JULIA\sbquarto_2\document.qmd:16
Error showing value of type Plots.Plot{Plots.GRBackend}
only png or svg allowed. got: :pdf
Stacktrace:
 [1] error(s::String)
   @ Base .\error.jl:35
 [2] _show(io::IOContext{IOBuffer}, ::MIME{Symbol("text/html")}, plt::Plots.Plot{Plots.GRBackend})
   @ Plots C:\Users\sbacelar\.julia\packages\Plots\ju9dp\src\output.jl:203
 [3] #invokelatest#2
   @ .\essentials.jl:892 [inlined]
 [4] invokelatest
   @ .\essentials.jl:889 [inlined]
 [5] show(io::IOContext{IOBuffer}, m::MIME{Symbol("text/html")}, plt::Plots.Plot{Plots.GRBackend})
   @ Plots C:\Users\sbacelar\.julia\packages\Plots\ju9dp\src\output.jl:232
 [6] show(io::IOContext{IOBuffer}, m::String, x::Plots.Plot{Plots.GRBackend})
   @ Base.Multimedia .\multimedia.jl:123

ERROR: Internal julia server error

Stack trace:
    at writeJuliaCommand (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:41672:19)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async executeJulia (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:41581:22)
    at async Object.execute (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:41332:20)
    at async renderExecute (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:77870:27)
    at async renderFileInternal (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:78038:43)
    at async renderFiles (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:77906:17)
    at async render (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:82734:21)
    at async Command.actionHandler (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:82882:32)
    at async Command.execute (file:///C:/Users/sbacelar/AppData/Local/Programs/Quarto/bin/quarto.js:8015:13)


## Windows 11 jupyter engine: quarto preview --to pdf

FATAL (C:/Users/sbacelar/AppData/Local/Programs/Quarto/share/filters/main.lua:7144) An error occurred:
Could not convert a SVG to a PDF for output. Please ensure that rsvg-convert is available on the path.
Error running filter C:/Users/sbacelar/AppData/Local/Programs/Quarto/share/filters/main.lua:
...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:2617: attempt to call a nil value (global 'crash_with_stack_trace')
stack traceback:
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:2196: in function 'fail'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:7144: in upvalue 'call_rsvg_convert'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:7168: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:7150>
        (...tail calls...)
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:2925: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:2919>
        [C]: in ?
        [C]: in method 'walk'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:535: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:524>
        (...tail calls...)
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1312: in local 'callback'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1330: in upvalue 'run_emulated_filter_chain'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1366: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1363>
stack traceback:
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:535: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:524>
        (...tail calls...)
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1312: in local 'callback'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1330: in upvalue 'run_emulated_filter_chain'
        ...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1366: in function <...lar/AppData/Local/Programs/Quarto/share/filters/main.lua:1363>

## WSL2 with Julia engine: quarto preview --to pdf

ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: /home/sbacelar/JULIA/sbquarto/document2.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ /home/sbacelar/JULIA/sbquarto/document2.qmd:16
Error showing value of type Plots.Plot{Plots.GRBackend}
only png or svg allowed. got: :pdf
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] _show(io::IOContext{IOBuffer}, ::MIME{Symbol("text/html")}, plt::Plots.Plot{Plots.GRBackend})
   @ Plots ~/.julia/packages/Plots/ju9dp/src/output.jl:203
 [3] #invokelatest#2
   @ ./essentials.jl:892 [inlined]
 [4] invokelatest
   @ ./essentials.jl:889 [inlined]
 [5] show(io::IOContext{IOBuffer}, m::MIME{Symbol("text/html")}, plt::Plots.Plot{Plots.GRBackend})
   @ Plots ~/.julia/packages/Plots/ju9dp/src/output.jl:232
 [6] show(io::IOContext{IOBuffer}, m::String, x::Plots.Plot{Plots.GRBackend})
   @ Base.Multimedia ./multimedia.jl:123

ERROR: Internal julia server error

Stack trace:
    at writeJuliaCommand (file:///opt/quarto/bin/quarto.js:41672:19)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async executeJulia (file:///opt/quarto/bin/quarto.js:41581:22)
    at async Object.execute (file:///opt/quarto/bin/quarto.js:41332:20)
    at async renderExecute (file:///opt/quarto/bin/quarto.js:77870:27)
    at async renderFileInternal (file:///opt/quarto/bin/quarto.js:78038:43)
    at async renderFiles (file:///opt/quarto/bin/quarto.js:77906:17)
    at async render (file:///opt/quarto/bin/quarto.js:82734:21)
    at async renderForPreview (file:///opt/quarto/bin/quarto.js:83760:26)
    at async render (file:///opt/quarto/bin/quarto.js:83643:29)

**Reply by cscheid** [reply]
@sbacelar please create a new discussion with this, you're commenting deep inside a separate thread.

**Comment by dmi3kno** [comment]
Do I understand correctly that Quarto uses its own "private" environment only to run QuartoNotebookRunner.jl? Is it absolutely certain that it will never require anything to be installed in that private environment directly? 

Second question: is there a reason why this setting is not default in Quarto?
```yml
engine: julia
julia:
  exeflags: ["--project=."]
```

**Reply by cderv** [reply]
Since this discussion, we tried to update the doc at https://quarto.org/docs/computations/julia.html#using-the-julia-engine to make it more clear. 

> Do I understand correctly that Quarto uses its own "private" environment only to run QuartoNotebookRunner.jl? Is it absolutely certain that it will never require anything to be installed in that private environment directly?

Yes. AFAIU this is use to start the Julia server to run the Julia cells. 

> Second question: is there a reason why this setting is not default in Quarto?
> 
> ```yaml
> engine: julia
> julia:
>   exeflags: ["--project=."]
> ```

The default used is `--project=@.` which seems to be more "clever". From our doc: 
> This makes Julia search for an environment (a Project.toml or JuliaProject.toml file) starting in the directory where the quarto notebook is stored and walking up the directory tree from there.

From Julia's doc: https://docs.julialang.org/en/v1/manual/environment-variables/index.html#JULIA_PROJECT
> If the variable is set to @. (note the trailing dot) then Julia tries to find a project directory that contains Project.toml or JuliaProject.toml file from the current directory and its parents. See also the chapter on [Code Loading](https://docs.julialang.org/en/v1/manual/code-loading/#code-loading).

Setting `--project=.` only consider the current directory IIUC. 

Is the current default not working ? It should cover the same usage case than `--project=.` 

Direct Julia engine support is quite new, so any feedback welcome really to improve it ! 

Hopefully the new doc helps understand;
